File: FamHelpDeskBackend/models/family_notification_settings.py
from models.base import FamHelpDeskBaseModel
from pynamodb.attributes import UnicodeAttribute, BooleanAttribute, NumberAttribute
from pynamodb.indexes import GlobalSecondaryIndex, AllProjection
from enum import Enum


class FamliyNotificationType(Enum):
    # Family
    NEW_FAMILY_CREATION = "New Family Created"

    # Family Membership
    FAMILY_MEMBERSHIP_APPROVED = "Family Membership Approved"
    FAMILY_MEMBERSHIP_DENIED = "Family Membership Denied"
    FAMILY_MEMBERSHIP_INVITATION = "Family Membership Invitation"
    FAMILY_MEMBER_JOINED = "Family Membership Accepted"
    FAMILY_MEMBERSHIP_LEFT = "Family Member left"
    FAMILY_MEMBERSHIP_REQUEST = "Family Membership Request"
    NEW_FAMILY_MEMEBER = "New Family Member"
    WELCOME_TO_FAMILY = "Welcome to Family"

    # Group Membership
    GROUP_MEMBERSHIP_APPROVED = "Group Membership Approved"
    GROUP_MEMBERSHIP_DENIED = "Group Membership Denied"
    GROUP_MEMBERSHIP_INVITATION = "Group Membership Invitation"
    GROUP_MEMBER_JOINED = "Group Membership Accepted"
    GROUP_MEMBERSHIP_LEFT = "Group Member left"
    GROUP_MEMBERSHIP_REQUEST = "Group Membership Request"

    # Ticket
    TICKET_CREATION_FAMILY = "Family Ticket Creation"
    TICKET_CREATION_GROUP = "Group Ticket Creation"
    TICKET_ASSIGNED = "Ticket Assigned"
    TICKET_COMMENT = "Ticket Comment"
    TICKET_STATUS_CHANGED = "Ticket Status Changed"


class FamilySettingsIndex(GlobalSecondaryIndex):
    """GSI for querying all settings for users in a family"""

    class Meta:
        index_name = "FamilySettingsIndex"
        projection = AllProjection()

    family_id = UnicodeAttribute(hash_key=True)
    user_id = UnicodeAttribute(range_key=True)


class FamilyNotificationSettings(FamHelpDeskBaseModel):
    """
    Per-family notification settings for users.

    PK: USER_PROFILE#{user_id}
    SK: NOTIFICATION_SETTINGS#{family_id}
    """

    user_id = UnicodeAttribute()
    family_id = UnicodeAttribute()

    # Notification type preferences with defaults

    # Family Membership
    family_membership_approved = BooleanAttribute(default=True)
    family_membership_denied = BooleanAttribute(default=False)
    family_membership_invitation = BooleanAttribute(default=False)
    family_membership_joined = BooleanAttribute(default=True)
    family_membership_left = BooleanAttribute(defaults=True)
    family_membership_request = BooleanAttribute(defaults=True)

    # Group Membership
    group_membership_approved = BooleanAttribute(defaults=False)
    group_membership_denied = BooleanAttribute(default=False)
    group_membership_invitation = BooleanAttribute(default=False)
    group_membership_joined = BooleanAttribute(default=False)
    group_membership_left = BooleanAttribute(default=False)
    group_membership_request = BooleanAttribute(default=False)

    # Tickets
    ticket_creation_family = BooleanAttribute(default=False)
    ticket_creation_group = BooleanAttribute(default=True)
    ticket_assigned = BooleanAttribute(default=True)
    ticket_comment = BooleanAttribute(default=True)
    ticket_satatus_change = BooleanAttribute(default=True)

    # Standard timestamp fields
    created_date = NumberAttribute()
    last_updated = NumberAttribute(null=True)

    # GSI
    family_settings_index = FamilySettingsIndex()

    @staticmethod
    def create_pk(user_id: str) -> str:
        """Create partition key for family notification settings"""
        return f"USER_PROFILE#{user_id}"

    @staticmethod
    def create_sk(family_id: str) -> str:
        """Create sort key for family notification settings"""
        return f"NOTIFICATION_SETTINGS#{family_id}"

    @staticmethod
    def clean_returned_settings(settings: "FamilyNotificationSettings") -> dict:
        """Clean and return family notification settings data"""
        return {
            "user_id": settings.user_id,
            "family_id": settings.family_id,
            "created_date": settings.created_date,
            "last_updated": settings.last_updated,
            # Faimly
            "family_membership_approved": settings.family_membership_approved,
            "family_membership_denied": settings.family_membership_denied,
            "family_membership_invitation": settings.family_membership_invitation,
            "family_membership_joined": settings.family_membership_joined,
            "family_membership_left": settings.family_membership_left,
            "family_membership_request": settings.family_membership_request,
            # Group Membership
            "group_membership_approved": settings.group_membership_approved,
            "group_membership_denied": settings.group_membership_denied,
            "group_membership_invitation": settings.group_membership_invitation,
            "group_membership_joined": settings.group_membership_joined,
            "group_membership_left": settings.group_membership_left,
            "group_membership_request": settings.group_membership_request,
            # Tickets
            "ticket_creation_family": settings.ticket_creation_family,
            "ticket_creation_group": settings.ticket_creation_group,
            "ticket_assigned": settings.ticket_assigned,
            "ticket_comment": settings.ticket_comment,
            "ticket_satatus_change": settings.ticket_satatus_change,
        }

    def save(self, **kwargs):
        """Override save to populate GSI attributes"""
        # GSI attributes are already set as model attributes
        # PynamoDB automatically handles GSI projection
        super().save(**kwargs)

File: FamHelpDeskiOS/FamHelpDesk/FamHelpDesk/Networking/Models/FamilyNotificationSettings.swift
import Foundation

struct FamilyNotificationSettingsResponse: Codable {
    let settings: FamilyNotificationSettings
}

struct FamilyNotificationSettings: Codable, Identifiable {
    var id: String { "\(userId)_\(familyId)" }

    let userId: String
    let familyId: String
    let createdDate: Int
    let lastUpdated: Int?

    // Family
    let newFamilyCreationEnabled: Bool
    let welcomeEnabled: Bool

    // Family Membership
    let welcomeToFamilyEnabled: Bool
    let newFamilyMemberEnabled: Bool
    let familyMembershipApproved: Bool
    let familyMembershipDenied: Bool
    let familyMembershipInvitation: Bool
    let familyMembershipJoined: Bool
    let familyMembershipLeft: Bool
    let familyMembershipRequest: Bool

    // Group Membership
    let groupMembershipApproved: Bool
    let groupMembershipDenied: Bool
    let groupMembershipAdded: Bool
    let groupMembershipJoined: Bool
    let groupMembershipLeft: Bool
    let groupMembershipRequest: Bool
    let newGroupCreation: Bool

    // Tickets
    let ticketCreationFamily: Bool
    let ticketCreationGroup: Bool
    let ticketAssigned: Bool
    let ticketComment: Bool
    let ticketStatusChange: Bool
    let ticketResolved: Bool

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case familyId = "family_id"
        case createdDate = "created_date"
        case lastUpdated = "last_updated"

        // Family
        case newFamilyCreationEnabled = "new_family_creation_enabled"
        case welcomeEnabled = "welcome_enabled"

        // Family Membership
        case welcomeToFamilyEnabled = "welcome_to_family_enabled"
        case newFamilyMemberEnabled = "new_family_member_enabled"
        case familyMembershipApproved = "family_membership_approved"
        case familyMembershipDenied = "family_membership_denied"
        case familyMembershipInvitation = "family_membership_invitation"
        case familyMembershipJoined = "family_membership_joined"
        case familyMembershipLeft = "family_membership_left"
        case familyMembershipRequest = "family_membership_request"

        // Group Membership
        case groupMembershipApproved = "group_membership_approved"
        case groupMembershipDenied = "group_membership_denied"
        case groupMembershipAdded = "group_membership_added"
        case groupMembershipJoined = "group_membership_joined"
        case groupMembershipLeft = "group_membership_left"
        case groupMembershipRequest = "group_membership_request"
        case newGroupCreation = "new_group_creation"

        // Tickets
        case ticketCreationFamily = "ticket_creation_family"
        case ticketCreationGroup = "ticket_creation_group"
        case ticketAssigned = "ticket_assigned"
        case ticketComment = "ticket_comment"
        case ticketStatusChange = "ticket_status_change"
        case ticketResolved = "ticket_resolved"
    }

    // Custom decoder that ignores unknown fields
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        userId = try container.decode(String.self, forKey: .userId)
        familyId = try container.decode(String.self, forKey: .familyId)
        createdDate = try container.decode(Int.self, forKey: .createdDate)
        lastUpdated = try container.decodeIfPresent(Int.self, forKey: .lastUpdated)

        // Family
        newFamilyCreationEnabled = try container.decode(Bool.self, forKey: .newFamilyCreationEnabled)
        welcomeEnabled = try container.decode(Bool.self, forKey: .welcomeEnabled)

        // Family Membership
        welcomeToFamilyEnabled = try container.decode(Bool.self, forKey: .welcomeToFamilyEnabled)
        newFamilyMemberEnabled = try container.decode(Bool.self, forKey: .newFamilyMemberEnabled)
        familyMembershipApproved = try container.decode(Bool.self, forKey: .familyMembershipApproved)
        familyMembershipDenied = try container.decode(Bool.self, forKey: .familyMembershipDenied)
        familyMembershipInvitation = try container.decode(Bool.self, forKey: .familyMembershipInvitation)
        familyMembershipJoined = try container.decode(Bool.self, forKey: .familyMembershipJoined)
        familyMembershipLeft = try container.decode(Bool.self, forKey: .familyMembershipLeft)
        familyMembershipRequest = try container.decode(Bool.self, forKey: .familyMembershipRequest)

        // Group Membership
        groupMembershipApproved = try container.decode(Bool.self, forKey: .groupMembershipApproved)
        groupMembershipDenied = try container.decode(Bool.self, forKey: .groupMembershipDenied)
        groupMembershipAdded = try container.decode(Bool.self, forKey: .groupMembershipAdded)
        groupMembershipJoined = try container.decode(Bool.self, forKey: .groupMembershipJoined)
        groupMembershipLeft = try container.decode(Bool.self, forKey: .groupMembershipLeft)
        groupMembershipRequest = try container.decode(Bool.self, forKey: .groupMembershipRequest)
        newGroupCreation = try container.decode(Bool.self, forKey: .newGroupCreation)

        // Tickets
        ticketCreationFamily = try container.decode(Bool.self, forKey: .ticketCreationFamily)
        ticketCreationGroup = try container.decode(Bool.self, forKey: .ticketCreationGroup)
        ticketAssigned = try container.decode(Bool.self, forKey: .ticketAssigned)
        ticketComment = try container.decode(Bool.self, forKey: .ticketComment)
        ticketStatusChange = try container.decode(Bool.self, forKey: .ticketStatusChange)
        ticketResolved = try container.decode(Bool.self, forKey: .ticketResolved)
    }
}

struct UpdateFamilyNotificationSettingsRequest: Codable {
    // Family
    let newFamilyCreationEnabled: Bool?
    let welcomeEnabled: Bool?

    // Family Membership
    let welcomeToFamilyEnabled: Bool?
    let newFamilyMemberEnabled: Bool?
    let familyMembershipApproved: Bool?
    let familyMembershipDenied: Bool?
    let familyMembershipInvitation: Bool?
    let familyMembershipJoined: Bool?
    let familyMembershipLeft: Bool?
    let familyMembershipRequest: Bool?

    // Group Membership
    let groupMembershipApproved: Bool?
    let groupMembershipDenied: Bool?
    let groupMembershipAdded: Bool?
    let groupMembershipJoined: Bool?
    let groupMembershipLeft: Bool?
    let groupMembershipRequest: Bool?
    let newGroupCreation: Bool?

    // Tickets
    let ticketCreationFamily: Bool?
    let ticketCreationGroup: Bool?
    let ticketAssigned: Bool?
    let ticketComment: Bool?
    let ticketStatusChange: Bool?
    let ticketResolved: Bool?

    enum CodingKeys: String, CodingKey {
        // Family
        case newFamilyCreationEnabled = "new_family_creation_enabled"
        case welcomeEnabled = "welcome_enabled"

        // Family Membership
        case welcomeToFamilyEnabled = "welcome_to_family_enabled"
        case newFamilyMemberEnabled = "new_family_member_enabled"
        case familyMembershipApproved = "family_membership_approved"
        case familyMembershipDenied = "family_membership_denied"
        case familyMembershipInvitation = "family_membership_invitation"
        case familyMembershipJoined = "family_membership_joined"
        case familyMembershipLeft = "family_membership_left"
        case familyMembershipRequest = "family_membership_request"

        // Group Membership
        case groupMembershipApproved = "group_membership_approved"
        case groupMembershipDenied = "group_membership_denied"
        case groupMembershipAdded = "group_membership_added"
        case groupMembershipJoined = "group_membership_joined"
        case groupMembershipLeft = "group_membership_left"
        case groupMembershipRequest = "group_membership_request"
        case newGroupCreation = "new_group_creation"

        // Tickets
        case ticketCreationFamily = "ticket_creation_family"
        case ticketCreationGroup = "ticket_creation_group"
        case ticketAssigned = "ticket_assigned"
        case ticketComment = "ticket_comment"
        case ticketStatusChange = "ticket_status_change"
        case ticketResolved = "ticket_resolved"
    }
}
